File: RotateAndCut.py
n_test_cases = int(input())
for _ in range(n_test_cases):
    n_rotates, line = map(str, input().split())
    n_rotates = int(n_rotates)
    start_ind = 0
    end_ind = len(line)
    cur_size = len(line)
    for i in range(min(24, n_rotates)):
        remove = cur_size // 4
        if i % 2 == 0:
            start_ind += remove
        else:
            end_ind -= remove
        cur_size -= remove
    print(line[start_ind:end_ind])

File: RotateAndCut.cpp
#include <iostream>
#include "bits/stdc++.h"

using namespace std;

int main() {
    int n;
    cin >> n;
    while (n--) {
        int nRotates, remove, curSize, startInd = 0, endInd;
        string line;
        cin >> nRotates >> line;
        endInd = curSize = line.length();
        for (int i = 0; i < min(nRotates, 24); ++i) {
            remove = curSize / 4;
            if (i % 2 == 0) {
                startInd += remove;
            }
            else {
                endInd -= remove;
            }
            curSize -= remove;
        }
        cout << line.substr(startInd, endInd - startInd) << endl;
    }
}